app/src/androidTest/java/org/wikipedia/model/SearchView.kt

```diff
@@ -0,0 +1,85 @@
+package org.wikipedia.model
+
+import android.view.View
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.UiController
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.ViewInteraction
+import androidx.test.espresso.action.ViewActions
+import androidx.test.espresso.assertion.ViewAssertions
+import androidx.test.espresso.contrib.RecyclerViewActions
+import androidx.test.espresso.matcher.BoundedMatcher
+import androidx.test.espresso.matcher.ViewMatchers
+import androidx.test.espresso.matcher.ViewMatchers.isAssignableFrom
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.withId
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+import org.junit.Assert
+import org.wikipedia.R
+import org.wikipedia.TestUtil
+
+class SearchView {
+    private val searchBarHint = "Search Wikipedia"
+    private val searchResultList = R.id.search_results_list
+    private val emptySearchResultsMessage = R.id.results_text
+
+    fun typeIntoSearchBar(text: String) : SearchView {
+        onView(ViewMatchers.withHint(searchBarHint))
+            .perform(ViewActions.typeText (text))
+
+        //todo use IdlingResources instead
+        TestUtil.delay(2)
+        return this
+    }
+
+    fun clickOnNthResult(itemIndex: Int): ArticleView {
+        onView(
+            Matchers.allOf(
+                ViewMatchers.withId(searchResultList),
+                ViewMatchers.isDisplayed()
+            )).perform(RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(itemIndex, ViewActions.click()))
+        return ArticleView()
+    }
+
+    fun resultsListShouldBeEmpty() {
+        Assert.assertEquals("No results", getText(onView(withId(emptySearchResultsMessage))))
+//        onView(recyclerViewSizeMatcher(0)).check(ViewAssertions.matches(isDisplayed()))
+    }
+
+    private fun getText(matcher: ViewInteraction): String {
```

Comment:
W wymaganiach napisane jest, żeby do asercji używać raczej metod jUnitowych, nie Espressowych. To sprawia, że jeśli np chcę pobrać tekst z elementu, muszę uciekać się do takich customowych funkcji. Czy to dobry trop (pewnie mogę zamienić to w extension method), czy błądzę, bo espresso jednak to jakoś ułatwia?

Reply:
Tak, tutaj trzeba niestety ale użyć własnych funkcji - najlepszym miejscem dla niej jest klasa bazowa dla page objectów np. `BasePage` - wtedy w każdym dziedziczącym po niej page object jesteś w stanie tego reużyć podając jako parametr matcher do widoku z którego chciałbym pobrać tekst. Jedyne co bym na szybko zmienił to w metodzie masz parametr `matcher` więc oczekiwałbym typu `ViewMatcher` a jest `ViewInteraction` ;)